Add scalability chapter (#16)
tikv/deep-dive-tikv: src/scalability/data-sharding.md

# Data sharding

## What is the partition

> For fault tolerance, TiKV replicates data to multiple nodes via Raft Consensus. For large datasets or high query throughput, having all the data is managed by one Raft state machine is not efficient and flexible, so we need to break the data up into partitions, also known as *sharding*, in TiKV a partition is called a `Region`.

Reviewer: ```suggestion
For fault tolerance, TiKV replicates data to multiple nodes via the Raft consensus algorithm. For large datasets or high query throughput, managing all the data by one Raft state machine is not efficient and flexible, so we need to break the data up into partitions, also known as *sharding*. In TiKV, a partition is called a `Region`.
```

> For fault tolerance, TiKV replicates multiple copies of the same data on different nodes based on Raft. But for that large dataset or high query throughput, all data is managed by one Raft state machine is not efficient and flexible, so we need to break the data up into partitions, also known as `sharding`. And we call a partition here is called a `Region` in TiKV.

Reviewer 2: ```suggestion
For fault tolerance, TiKV replicates data to multiple nodes via Raft Consensus. For large datasets or high query throughput, having all data is managed by one Raft state machine is not efficient and flexible, so we need to break the data up into partitions, also known as *sharding*. And we call a partition here is called a `Region` in TiKV.
```

> With partitioning, we can move partitions to balance the load among all nodes of a cluster. For how to decide which partitions to store on which nodes isn't covered here. It will be discussed later in chapter related to PD.

Reviewer 2: ```suggestion
With partitioning, we can move partitions to balance the load among all nodes of a cluster. How to decide which partitions to store on which nodes isn't covered here, it will be discussed later in chapter related to PD.
```

## Partitioning

> Different partitions can be placed on different nodes, thus a large dataset can be distributed across many disks, and the query load can be distributed across many processors. But partitioning suffers from the issue of *hot spots*, where high load may end up on one partition but leave the other nodes idle. To avoid that, we can simply assign records to nodes randomly, but when reading a particular item there is no fast way of knowing which node is on. 

Reviewer: ```suggestion
Different partitions can be placed on different nodes, and thus a large dataset can be distributed across many disks, and the query load can be distributed across many processors. But partitioning suffers from the issue of *hot spots*, where high load may end up on one partition but leave the other nodes idle. To avoid that, we can simply assign records to nodes randomly, but when reading a particular item there is no fast way of knowing which node is on. 
```

Reviewer: Note: "thus" is an adverb rather than a conjunction.

> Different partitions can be placed on different nodes, thus a large dataset can be distributed across many disks, and the query load can be distributed across many processors. But partitioning suffers from hot spot issue that high load may end up on one partition which makes only one node busy but other nodes idle. To avoid that, we can simply assign records to nodes randomly, but when reading a particular item there is no fast way of knowing which node is on. 

Reviewer 2: ```suggestion
Different partitions can be placed on different nodes, thus a large dataset can be distributed across many disks, and the query load can be distributed across many processors. But partitioning suffers from the issue of *hot spots*, where high load may end up on one partition but leave the other nodes idle. To avoid that, we can simply assign records to nodes randomly, but when reading a particular item there is no fast way of knowing which node is on. 
```

> There two general ways to map keys to partitions:

Reviewer: ```suggestion
There are two general ways to map keys to partitions:
```

> TiKV uses simple key-value data model, here are two general ways to map keys to partitions:

Reviewer 2: ```suggestion
There are two general ways to map keys to partitions:
```

### Partitioning by hash of key

> One way of partitioning is to assign a continuous range of hashes of keys to each partition. The partition boundaries can be evenly spaced, or they can be chosen by some kind of algorithm like [consistent hashing](https://en.wikipedia.org/wiki/Consistent_hashing). Due to the hash, keys can be distributed fairly among the partitions, effectively reducing hot spots, though not totally avoiding them. Unfortunately, with this method unable to do efficient range queries compared to partitioning by key range.

Reviewer: ```suggestion
One way of partitioning is to assign a continuous range of hashes of keys to each partition. The partition boundaries can be evenly spaced, or they can be chosen by some kind of algorithm like [consistent hashing](https://en.wikipedia.org/wiki/Consistent_hashing). Due to the hash, keys can be distributed fairly among the partitions, effectively reducing hot spots, though not totally avoiding them. Unfortunately, this method fails to do efficient range queries, compared to partitioning by key range.
```

> One way of partitioning is to assign a continuous range of hashes of keys to each partition. The partition boundaries can be evenly spaced, or they can be chosen by some kind of algorithm like [consistent hashing](https://en.wikipedia.org/wiki/Consistent_hashing). Due to hash, it can distribute keys fairly among the partitions which effectively reduces hot spots though not totally avoids. Whereas, it is unable to do efficient range queries comparing with partitioning by key range.

Reviewer 2: ```suggestion
One way of partitioning is to assign a continuous range of hashes of keys to each partition. The partition boundaries can be evenly spaced, or they can be chosen by some kind of algorithm like [consistent hashing](https://en.wikipedia.org/wiki/Consistent_hashing). Due the hash, keys can be distributed fairly among the partitions, effectively reducing hot spots, though not totally avoiding them. Unfortunately, with this method we are unable to do efficient range queries compared to partitioning by key range.
```

### Partitioning by key range

> The other way of partitioning is to assign a continuous range of keys to each partition. It needs to record the boundaries between the ranges to determine which partition contains a given key, then requests can be sent to the appropriate node. One key feature of key range partition is that it is friendly to range scans. However, the downside of the sorted order is that sequential reads and writes can lead to hot spots. 

Reviewer: ```suggestion
The other way of partitioning is to assign a continuous range of keys to each partition. It needs to record the boundaries between the ranges to determine which partition contains a given key, and then requests can be sent to the appropriate node. One key feature of key range partition is that it is friendly to range scans. However, the downside of the sorted order is that sequential reads and writes can lead to hot spots. 
```

> The other way of partitioning is to assign a continuous range of keys to each partition. It needs to record the boundaries between the ranges to determine which partition contains a given key, then requests can be sent to the appropriate node. One key feature of key range partition is friendly to range scans. However, the downside of the sorted order is that sequential reads and writes can lead to hot spots. 

Reviewer 2: ```suggestion
The other way of partitioning is to assign a continuous range of keys to each partition. It needs to record the boundaries between the ranges to determine which partition contains a given key, then requests can be sent to the appropriate node. One key feature of key range partition is that it is friendly to range scans. However, the downside of the sorted order is that sequential reads and writes can lead to hot spots. 
```

> Partitioning by range is the approach that TiKV uses. The main reason to choose it is scan-friendliness, meanwhile, for the split or merge of regions, TiKV only needs to change meta-information about the range of the Region which avoids moving actual data in a large extent.

Reviewer: ```suggestion
Partitioning by range is the approach that TiKV uses. The main reason to choose it is scan-friendliness, meanwhile, for the split or merge of Regions, TiKV only needs to change meta-information about the range of the Region, which avoids moving actual data in a large extent.
```

> And it is the partitioning approach that TiKV uses. The main reason to choose it is scan-friendly, meanwhile, for the split or merge of Region, TiKV only needs to change meta-information about the range of the Region which avoids moving actual data in a large extent.

Reviewer 2: Try to avoid starting sentences with "And".
```suggestion
Partitioning by range is the approach that TiKV uses. The main reason to choose it is scan-friendliness, meanwhile, for the split or merge of regions, TiKV only needs to change meta-information about the range of the Region which avoids moving actual data in a large extent.
```
